### PR TITLE
Bind this.onError for XHR requests in TurboGraft.Remote

### PR DESCRIPTION
**Pull request by davidcornu**
I have a feeling this is causing https://bugsnag.com/shopify/shopify-js/errors/5666f3c78f2356495b6fe53f?filters%5Bevent.since%5D%5B%5D=30d.

r @qq99 
cc @arapehl

### CODE REVIEW

File: lib/assets/javascripts/turbograft/remote.coffee [hunk 1]
```diff
@@ -33,7 +33,7 @@ class TurboGraft.Remote
       triggerEventFor 'turbograft:remote:start', @initiator,
         xhr: xhr
 
-    xhr.addEventListener 'error', @onError
+    xhr.addEventListener 'error', @onError.bind(@)
```
Comment (qq99) [inline lib/assets/javascripts/turbograft/remote.coffee]: can you do `.bind(this)`?

**Comment by qq99** [comment]
Looks like it might address a lot of https://github.com/Shopify/turbograft/issues/118

cc @Shopify/tnt

**Comment by davidcornu** [comment]
By the looks of things, the `success` and `fail` options of `TurboGraft.Remote` don't have any test coverage. I'll add some.

File: lib/assets/javascripts/turbograft/remote.coffee [hunk 2]
```diff
@@ -33,7 +33,7 @@ class TurboGraft.Remote
       triggerEventFor 'turbograft:remote:start', @initiator,
         xhr: xhr
 
-    xhr.addEventListener 'error', @onError
+    xhr.addEventListener 'error', @onError.bind(this)
```
Comment (lemonmade) [inline lib/assets/javascripts/turbograft/remote.coffee]: More CoffeeScript-ey solution is to use `=>` in the method definition (which binds the method in the constructor).

File: lib/assets/javascripts/turbograft/remote.coffee [hunk 3]
```diff
@@ -116,7 +116,7 @@ class TurboGraft.Remote
       enabledInputs.push(input)
     enabledInputs
 
-  onSuccess: (ev) ->
+  onSuccess: (ev) =>
```
Comment (davidcornu) [inline lib/assets/javascripts/turbograft/remote.coffee]: Figured I'd switch this to a fat arrow too seen as it's always used as a callback

File: test/javascripts/remote_test.coffee [hunk 4]
```diff
@@ -67,6 +67,53 @@ describe 'Remote', ->
...
+
+      remote = new TurboGraft.Remote
+        httpRequestType: "POST"
+        httpUrl: "/foo/bar"
+        refreshOnError: "foo"
+        fail: done
+      , @initiating_target
+
+      simulateError = ->
+        listener(target: remote.xhr) for listener in remote.xhr.eventListeners.error
```
Comment (davidcornu) [inline test/javascripts/remote_test.coffee]: This was the only way I could find to simulate a `xhr.onError` event

**Comment by davidcornu** [comment]
@qq99 & @lemonmade - round 2 please 😁

File: test/javascripts/remote_test.coffee [hunk 5]
```diff
@@ -67,6 +67,53 @@ describe 'Remote', ->
...
+    it 'will call options.fail() on HTTP failures', (done) ->
+      server = sinon.fakeServer.create()
+      server.respondWith("POST", "/foo/bar",
+            [422, { "Content-Type": "text/html" },
+             '<div id="foo" refresh="foo">Error occured</div>'])
+      remote = new TurboGraft.Remote
+        httpRequestType: "POST"
+        httpUrl: "/foo/bar"
+        refreshOnError: "foo"
+        fail: done
```
Comment (lemonmade) [inline test/javascripts/remote_test.coffee]: If this never failed, would this test just timeout? Might be worth adding the success case that throws a meaningful error.

File: test/javascripts/remote_test.coffee (hunk 4 above)
Comment (lemonmade) [inline test/javascripts/remote_test.coffee]: Yeah, I can't really think of how to test this nicely either :S

File: test/javascripts/remote_test.coffee [hunk 6]
```diff
@@ -67,6 +67,53 @@ describe 'Remote', ->
...
+      , @initiating_target
+
+      simulateError = ->
+        listener(target: remote.xhr) for listener in remote.xhr.eventListeners.error
+
+      setTimeout(simulateError, 0)
```
Comment (lemonmade) [inline test/javascripts/remote_test.coffee]: Why do you have to delay this?

File: test/javascripts/remote_test.coffee [hunk 7]
```diff
@@ -67,6 +67,53 @@ describe 'Remote', ->
...
+
+    it 'will call options.success() on success', (done) ->
+      server = sinon.fakeServer.create()
+      server.respondWith("POST", "/foo/bar",
```
Comment (lemonmade) [inline test/javascripts/remote_test.coffee]: whitespace is a little wonky on these, I would probably do one arg per line.

File: test/javascripts/remote_test.coffee [hunk 8]
```diff
@@ -67,6 +67,53 @@ describe 'Remote', ->
...
+            [200, { "Content-Type": "text/html" },
+             '<div>Hey there</div>'])
+      remote = new TurboGraft.Remote
+        httpRequestType: "POST"
+        httpUrl: "/foo/bar"
+        refreshOnError: "foo"
+        success: done
```
Comment (lemonmade) [inline test/javascripts/remote_test.coffee]: SO I guess, if these tests don't time out that's the success case. In addition to the "adding an error for the fail case" suggestion I made above, maybe use a spy for the success case callback, and check that the spy was called (again, just to be more explicit)?

File: test/javascripts/remote_test.coffee (hunk 4 above)
Comment (qq99) [inline test/javascripts/remote_test.coffee]: Interesting

**Comment by qq99** [comment]
Just going back to look at https://github.com/Shopify/turbograft/commit/dafb3a7f2ee851b64a69c22086ddb6c3c9b9e00a#diff-58f1b55d4025fab95951247202574f55L156 to analyze some more...

Looks like previously, `@refreshOnError` would have been undefined in some cases, and so it'd just do a general `triggerEventFor` and that'd be the end of it. It also looks like the network-level failures were/are the ones with problem (lacking proper `this`), combined with trying to do multi-level object lookup now

@davidcornu LGTM let me know when you'd like me to 🎩 this against core

File: test/javascripts/remote_test.coffee (hunk 6 above)
Comment (davidcornu) [inline test/javascripts/remote_test.coffee]: To guarantee async. XHR error callbacks would never be called within the same tick.

**Comment by davidcornu** [comment]
> It also looks like the network-level failures were/are the ones with problem (lacking proper `this`), combined with trying to do multi-level object lookup now

Yeah that was my conclusion.

File: test/javascripts/remote_test.coffee (hunk 7 above)
Comment (davidcornu) [inline test/javascripts/remote_test.coffee]: @lemonmade was adhering to the style in  the rest of the file. Should I update all of it or just this one?

File: test/javascripts/remote_test.coffee (hunk 8 above)
Comment (davidcornu) [inline test/javascripts/remote_test.coffee]: The error scenario is a timeout for both of these cases. I can't think of a way to give a better error message though as there's no way to hook into the timeout behaviour. What I want is

```javascript
it('calls the callback within 2 seconds', function(done){
  var failureTimeout = setTimeout(function(){
    done(new Error("callback was not called"));
  }, 2000);

  doSomethingAsync(function(){
    clearTimeout(failureTimeout);
    done();
  });
});
```

but that seems like a lot of code to make the error message better.

File: test/javascripts/remote_test.coffee (hunk 7 above)
Comment (lemonmade) [inline test/javascripts/remote_test.coffee]: If it's the same everywhere, KKSO 👍

File: test/javascripts/remote_test.coffee (hunk 8 above)
Comment (lemonmade) [inline test/javascripts/remote_test.coffee]: talked on Slack, I'm good with the current version 👍

Comment (davidcornu) [inline test/javascripts/remote_test.coffee]: Talked to @lemonmade on Slack. Timeout is acceptable in this case.

**Comment by davidcornu** [comment]
@qq99 should be good to go. LMK if you want any help with 🎩 .

**Comment by qq99** [comment]
Tested it in a few scenarios (order actions, some 422s) and looks good 🎩

**Comment by qq99** [comment]
🚢

**Pull request merged**